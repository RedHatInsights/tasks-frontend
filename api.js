import axios from 'axios';
import {
  TASKS_API_ROOT,
  AVAILABLE_TASKS_ROOT,
  EXECUTED_TASK_ROOT,
  SYSTEMS_ROOT,
  JOB_ROOT,
} from './src/constants';

const returnErrOrData = (response) => {
  if (response.status === 200) {
    return response.data;
  } else {
    return response;
  }
};

const getTasks = async (path) => {
  const response = await axios
    .get(TASKS_API_ROOT.concat(path))
    .catch(function (error) {
      return error;
    });

  return returnErrOrData(response);
};

const postTask = async (path, data) => {
  const response = await axios
    .post(TASKS_API_ROOT.concat(path), data)
    .catch(function (error) {
      return error;
    });

  return returnErrOrData(response);
};

const deleteTask = async (path) => {
  const response = await axios
    .delete(TASKS_API_ROOT.concat(path))
    .catch(function (error) {
      return error;
    });

  return returnErrOrData(response);
};

export const fetchAvailableTasks = (path = '') => {
  return getTasks(AVAILABLE_TASKS_ROOT.concat(path));
};

export const fetchAvailableTask = (slug) => {
  return getTasks(AVAILABLE_TASKS_ROOT.concat(`/${slug}`));
};

export const fetchExecutedTasks = (path) => {
  if (path) {
    return getTasks(EXECUTED_TASK_ROOT.concat(path));
  } else {
    return getTasks(EXECUTED_TASK_ROOT);
  }
};

export const fetchExecutedTask = (id, jobs_path = '') => {
  let idPath = `/${id}${jobs_path}`;
  return getTasks(EXECUTED_TASK_ROOT.concat(idPath));
};

export const fetchExecutedTaskJobs = (id, path) => {
  return fetchExecutedTask(id, `/jobs${path}`);
};

export const fetchSystems = (path = '', slug = '') => {
  if (slug) {
    return getTasks(AVAILABLE_TASKS_ROOT.concat(`/${slug}/systems${path}`));
  } else {
    return getTasks(SYSTEMS_ROOT.concat(path));
  }
};

export const executeTask = (body) => {
  return postTask(EXECUTED_TASK_ROOT, body);
};

export const deleteExecutedTask = (id) => {
  return deleteTask(EXECUTED_TASK_ROOT.concat(`/${id}`));
};

export const cancelExecutedTask = (id) => {
  return postTask(EXECUTED_TASK_ROOT.concat(`/${id}/cancel`));
};

export const getLogs = (id) => {
  return getTasks(JOB_ROOT.concat(`/${id}/stdout`))
};
