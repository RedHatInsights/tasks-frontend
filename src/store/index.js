import { createContext } from 'react';
import { applyReducerHash } from '@redhat-cloud-services/frontend-components-utilities/ReducerRegistry';
import { getRegistry } from '@redhat-cloud-services/frontend-components-utilities/Registry';
import promiseMiddleware from 'redux-promise-middleware';
import { notificationsReducer } from '@redhat-cloud-services/frontend-components-notifications/redux';
import { globalFilterReducer } from '../modules/reducers';

let registry;

export const RegistryContext = createContext({
  getRegistry: () => {},
});

export function init(...middleware) {
  registry = getRegistry({}, [
    promiseMiddleware,
    ...middleware.filter((item) => typeof item !== 'undefined'),
  ]);

  registry.register({
    notifications: notificationsReducer,
    globalFilterState: globalFilterReducer,
  });

  return registry;
}

export function getStore() {
  return registry.getStore();
}

const selectRows = (rows, selected) =>
  rows.map((row) => ({
    ...row,
    selected: selected.includes(row.id),
  }));

export const entitiesReducer = () =>
  applyReducerHash({
    ['INVENTORY_INIT']: () => ({
      rows: [],
      total: 0,
    }),
    ['RESET_PAGE']: (state) => ({
      ...state,
      page: 1,
      perPage: 10,
    }),
    ['SELECT_ENTITIES']: (state, { payload: { selected } }) => {
      return {
        ...state,
        rows: selectRows(state.rows || [], selected),
      };
    },
  });
