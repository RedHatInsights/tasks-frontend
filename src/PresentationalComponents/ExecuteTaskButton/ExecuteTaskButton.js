import React from 'react';
import { Button } from '@patternfly/react-core';
import propTypes from 'prop-types';
import { executeTask } from '../../../api';

const ExecuteTaskButton = ({
  classname,
  ids,
  setExecuteTaskResult,
  slug,
  taskName,
  variant,
}) => {
  const buildApiBody = () => {
    return {
      task: slug,
      hosts: ids,
      name: taskName,
    };
  };

  const submitTask = async () => {
    let result = await executeTask(buildApiBody());
    setExecuteTaskResult(result);
  };

  return (
    <Button
      aria-label={`${slug}-submit-task-button`}
      className={classname}
      variant={variant}
      onClick={() => submitTask()}
      isDisabled={!ids?.length}
    >
      Execute task
    </Button>
  );
};

ExecuteTaskButton.propTypes = {
  classname: propTypes.string,
  ids: propTypes.array,
  setExecuteTaskResult: propTypes.func,
  slug: propTypes.string,
  taskName: propTypes.string,
  variant: propTypes.string,
};

export default ExecuteTaskButton;
