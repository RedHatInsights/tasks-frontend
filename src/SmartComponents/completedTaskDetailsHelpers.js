import { dispatchNotification } from '../Utilities/Dispatcher';
import { fetchExecutedTask, fetchExecutedTaskJobs } from '../../api';
import {
  JOB_FAILED_MESSAGE,
  JOB_RUNNING_MESSAGE,
  JOB_TIMED_OUT_MESSAGE,
} from '../constants';

export const getSelectedSystems = (completedTaskJobs) => {
  let systemIds = [];
  completedTaskJobs.forEach((job) => {
    if (job.display_name) {
      systemIds.push(job.system);
    }
  });

  return systemIds;
};

export const isError = (result) => {
  const statusString = result?.response?.status.toString();
  return result?.response?.status && statusString[0] !== '2';
};

export const createNotification = (result) => {
  dispatchNotification({
    variant: 'danger',
    title: 'Error',
    description: result.message,
    dismissable: true,
    autoDismiss: false,
  });
};

export const fetchTask = async (id, setError) => {
  let taskDetails = await fetchExecutedTask(id);

  if (isError(taskDetails)) {
    createNotification(taskDetails);
    setError(taskDetails);
  } else {
    return taskDetails;
  }
};

export const fetchTaskJobs = async (taskDetails, setError) => {
  let path = `?limit=1000&offset=0`;
  const taskJobs = await fetchExecutedTaskJobs(taskDetails.id, path);

  if (isError(taskJobs)) {
    createNotification(taskJobs);
    setError(taskJobs);
  } else {
    taskDetails.alerts_count =
      taskJobs.data.filter((item) => {
        if (item.status === 'Timeout' || item.status === 'Failure') {
          return item;
        } else {
          return item.results.alert;
        }
      }).length || '-';
    taskDetails.system_count = taskJobs.data.length;
    taskJobs.data.forEach((job) => {
      if (job.status === 'Failure') {
        job.results.message = JOB_FAILED_MESSAGE;
      } else if (job.status === 'Timeout') {
        job.results.message = JOB_TIMED_OUT_MESSAGE;
      } else if (job.status === 'Running') {
        job.results.message = JOB_RUNNING_MESSAGE;
      }
    });

    return taskJobs.data;
  }
};

export const hasDetails = (completedTaskJob) => {
  return completedTaskJob.status === 'Success' &&
    completedTaskJob.results.report_json?.entries?.length
    ? true
    : false;
};
