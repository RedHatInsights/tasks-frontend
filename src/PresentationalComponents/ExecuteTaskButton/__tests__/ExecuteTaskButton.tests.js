import React from 'react';
import { render, screen, waitFor } from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import { MemoryRouter } from 'react-router-dom';
import { Provider } from 'react-redux';
import { init } from '../../../store';

import ExecuteTaskButton from '../ExecuteTaskButton';
import { executeTask } from '../../../../api';
import { successResponse } from '../../../SmartComponents/__tests__/__fixtures__/completedTaskDetailsHelpers.fixtures';

jest.mock('../../../../api');

describe('ExecuteTaskButton', () => {
  let props;
  const store = init().getStore();

  beforeEach(() => {
    props = {
      classname: 'some-class-name',
      ids: ['abcd-1234'],
      setExecuteTaskResult: jest.fn(),
      slug: 'taska',
      taskName: 'Task A',
      variant: 'primary',
    };
  });

  it('should render correctly', async () => {
    const { asFragment } = render(<ExecuteTaskButton {...props} />);

    expect(asFragment()).toMatchSnapshot();
  });

  it('should execute on click', async () => {
    executeTask.mockImplementation(async () => {
      return {
        successResponse,
        data: { id: '1' },
      };
    });

    render(
      <MemoryRouter keyLength={0}>
        <Provider store={store}>
          <ExecuteTaskButton {...props} />
        </Provider>
      </MemoryRouter>
    );

    await waitFor(() =>
      userEvent.click(screen.getByLabelText('taska-submit-task-button'))
    );
    expect(props.setExecuteTaskResult).toHaveBeenCalled();
  });
});
