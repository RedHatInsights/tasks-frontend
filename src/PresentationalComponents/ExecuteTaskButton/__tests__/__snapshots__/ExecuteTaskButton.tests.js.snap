// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ExecuteTaskButton should render correctly 1`] = `
<DocumentFragment>
  <button
    aria-disabled="false"
    aria-label="taska-submit-task-button"
    class="pf-v5-c-button pf-m-primary some-class-name"
    data-ouia-component-id="OUIA-Generated-Button-primary-1"
    data-ouia-component-type="PF5/Button"
    data-ouia-safe="true"
    type="button"
  >
    Execute task
  </button>
</DocumentFragment>
`;
